qiskit/algorithms/gradients/param_shift_estimator_gradient.py

```diff
@@ -0,0 +1,150 @@
+
+from .base_estimator_gradient import BaseEstimatorGradient
+from .estimator_gradient_result import EstimatorGradientResult
+
+from .utils import (
+    make_param_shift_gradient_circuit_data,
+    make_param_shift_base_parameter_values,
+)
+
+
+class ParamShiftEstimatorGradient(BaseEstimatorGradient):
+    """Compute the gradients of the expectation values by the parameter shift rule"""
+
+    def __init__(self, estimator: BaseEstimator, **run_options):
+        """
+        Args:
+            estimator: The estimator used to compute the gradients.
+            run_options: Backend runtime options used for circuit execution. The order of priority is:
+                run_options in `run` method > gradient's default run_options > primitive's default
+                setting. Higher priority setting overrides lower priority setting.
+        """
+        self._gradient_circuit_data_dict = {}
+        self._base_parameter_values_dict = {}
+        super().__init__(estimator, **run_options)
+
+    def _run(
+        self,
+        circuits: Sequence[QuantumCircuit],
+        observables: Sequence[BaseOperator | PauliSumOp],
+        parameter_values: Sequence[Sequence[float]],
+        parameters: Sequence[Sequence[Parameter] | None] | None = None,
+        **run_options,
+    ) -> EstimatorGradientResult:
+        """Compute the estimator gradients on the given circuits."""
+        # if parameters is none, all parameters in each circuit are differentiated.
+        if parameters is None:
+            parameters = [None for _ in range(len(circuits))]
+
+        jobs, result_indices_all, coeffs_all = [], [], []
+        for circuit, observable, parameter_values_, parameters_ in zip(
+            circuits, observables, parameter_values, parameters
+        ):
+            # a set of parameters to be differentiated
+            if parameters_ is None:
+                param_set = set(circuit.parameters)
+            else:
+                param_set = set(parameters_)
+
+            gradient_circuit_data = self._gradient_circuit_data_dict.get(id(circuit))
+            base_parameter_values_all = self._base_parameter_values_dict.get(id(circuit))
+            if gradient_circuit_data is None and base_parameter_values_all is None:
+                gradient_circuit_data = make_param_shift_gradient_circuit_data(circuit)
+                self._gradient_circuit_data_dict[id(circuit)] = gradient_circuit_data
+                base_parameter_values_all = make_param_shift_base_parameter_values(
+                    gradient_circuit_data
+                )
+                self._base_parameter_values_dict[id(circuit)] = base_parameter_values_all
+
+            plus_offsets, minus_offsets = [], []
+            gradient_circuit = gradient_circuit_data.gradient_circuit
+            gradient_parameter_values = np.zeros(
+                len(gradient_circuit_data.gradient_circuit.parameters)
+            )
```

Comment:
There are very similar codes in sampler_gradient and estimator_gradient. It would be great if you can make a param_shift_utils.py, for example, to put the common part.